### PR TITLE
Use git user name for author name

### PR DESCRIPTION
**Pull request by sh4869**
If working directory is git repository and user doesn't set `--author` option, Use git user name for author name.

### CODE REVIEW

File: lib/src/cli_app.dart [hunk 1]
```diff
@@ -169,6 +169,10 @@ additional analytics to help us improve Stagehand [y/yes/no]? """);
 
     String author = options['author'];
 
+    if (author == '<your name>'  && io.FileSystemEntity.isDirectorySync(dir.path + "/.git")) {
+      author = io.Process.runSync('git', ['config', 'user.name']).stdout.trim();
```
Comment (sethladd) [inline lib/src/cli_app.dart]: What if this fails?

**Comment by sethladd** [comment]
I like the idea, thanks for the patch. Left some comments.

**Comment by sh4869** [comment]
I think if this fails, this should only outputs error info and don't exit because this is not main function of this tool. I write some code following this policy, please check.  Sorry for my poor English.

**Comment by sh4869** [comment]
cc @sethladd

File: lib/src/cli_app.dart [hunk 2]
```diff
@@ -169,6 +169,27 @@ additional analytics to help us improve Stagehand [y/yes/no]? """);
...
+        var stdout = io.Process.runSync('git', ['config', 'user.name']).stdout;
+        if (stdout != null) {
+          author = stdout.trim();
+          _out('Use Git user name for author name.');
+        } else {
+          _out('Error: git user name is not set.');
+        }
+      } catch (e, st) {
+        if (e is io.ProcessException) {
+          _out('Error: git is not on your path or fail to run.');
```
Comment (sethladd) [inline lib/src/cli_app.dart]: "or failed to run"

File: lib/src/cli_app.dart [hunk 3]
```diff
@@ -169,6 +169,27 @@ additional analytics to help us improve Stagehand [y/yes/no]? """);
 
     String author = options['author'];
 
+    if (author == '<your name>' &&
+        io.FileSystemEntity.isDirectorySync(dir.path + "/.git")) {
```
Comment (sethladd) [inline lib/src/cli_app.dart]: I think this needs to be unindented.

File: lib/src/cli_app.dart [hunk 4]
```diff
@@ -169,6 +169,27 @@ additional analytics to help us improve Stagehand [y/yes/no]? """);
...
+      _out(
+          "This directory is git repository.  Try to use git user name for author name ...");
```
Comment (sethladd) [inline lib/src/cli_app.dart]: Let's use user.name instead of "user name"

File: lib/src/cli_app.dart [hunk 5]
```diff
@@ -169,6 +169,27 @@ additional analytics to help us improve Stagehand [y/yes/no]? """);
...
+      try {
+        var stdout = io.Process.runSync('git', ['config', 'user.name']).stdout;
+        if (stdout != null) {
+          author = stdout.trim();
+          _out('Use Git user name for author name.');
+        } else {
+          _out('Error: git user name is not set.');
```
Comment (sethladd) [inline lib/src/cli_app.dart]: please use user.name instead of "user name"

File: lib/src/cli_app.dart [hunk 6]
```diff
@@ -169,6 +169,27 @@ additional analytics to help us improve Stagehand [y/yes/no]? """);
...
+        }
+      } catch (e, st) {
```
Comment (sethladd) [inline lib/src/cli_app.dart]: please use full variable names

**Comment by kwalrath** [comment]
@sethladd should I merge this?

**Comment by sethladd** [comment]
I haven't reviewed the code, but I like the intent of the patch. :)

**Comment by kwalrath** [comment]
@devoncarew can you review the code?

**Comment by devoncarew** [comment]
Hey @sh4869, I know you've had a lot of review on this small patch, so I apologize in advance for the last-minute review. Here are my thoughts:

- we don't want to display any errors to the user about being unable to run `git config`; it's probably not something they're directory expecting stagehand to do
- we want to be a bit more resilient if git returns an error code

I would suggest something like this:

```
if (!options.wasParsed('author')) {
  ProcessResult result = io.Process.runSync('git', ['config', 'user.name']);
  if (result.exitCode == 0) {
    author = result.stdout.trim();
  }
}
```

**Comment by kwalrath** [comment]
Also, @sh4869 could you please add yourself to the AUTHORS file?

**Comment by sh4869** [comment]
I fixed code and added name to the AUTHORS.

File: lib/src/cli_app.dart [hunk 7]
```diff
@@ -169,6 +169,14 @@ additional analytics to help us improve Stagehand [y/yes/no]? """);
 
     String author = options['author'];
 
+    if (!options.wasParsed('author') &&
+        io.FileSystemEntity.isDirectorySync(dir.path + "/.git")) {
+      try {
+        var stdout = io.Process.runSync('git', ['config', 'user.name']).stdout;
```
Comment (devoncarew) [inline lib/src/cli_app.dart]: This doesn't check the return code - `git config` could return a non-zero code, in which case we don't want to use it's output.

File: lib/src/cli_app.dart [hunk 8]
```diff
@@ -169,6 +169,14 @@ additional analytics to help us improve Stagehand [y/yes/no]? """);
 
     String author = options['author'];
 
+    if (!options.wasParsed('author') &&
+        io.FileSystemEntity.isDirectorySync(dir.path + "/.git")) {
```
Comment (devoncarew) [inline lib/src/cli_app.dart]: I'm not sure if this check is helpful. The `dir` directory will have just been created. We just want to try shelling out to git if the user didn't supply an author option.

Comment (sh4869) [inline lib/src/cli_app.dart]: I fixed it.

**Comment by sh4869** [comment]
ping

File: lib/src/cli_app.dart (hunk 8 above)
Comment (devoncarew) [inline lib/src/cli_app.dart]: Can you remove this portion of the check: `io.FileSystemEntity.isDirectorySync(dir.path + "/.git")`? We don't need to be conditional on the presence of this directory.

**Comment by kwalrath** [comment]
Sorry, @sh4869, I was out of town. @devoncarew, do you approve this PR?

**Comment by devoncarew** [comment]
Yes, lgtm. Thanks!

**Comment by kwalrath** [comment]
@sh4869 if you can resolve the conflicts, I'll merge this. Thank you!

**Comment by sh4869** [comment]
@kwalrath I did `git rebase`.

**Comment by kwalrath** [comment]
Thanks, @sh4869!

**Pull request closed without merging**